cfp: Added CFP for blocking traffic to certain endpoints with default allow (#12)
cilium/design-cfps: cilium/CFP-26135-default-egress-allow.md

This proposal aims to bring the management of cloud provider specific policies within Cilium. The primary goal of this proposal is to integrate the cloud provider-specific policies into the Cilium framework. By doing so, we aim to address potential conflicts and other concerns associated with managing these policies across multiple components.

## Goals

- Block egress traffic to certain endpoints without blocking all
- Behavior of existing network policy or cilium network policy should not be affected

## Non Goals
 - Ingress Restriction without blocking all
 - Applying this policy for specific endpoints
 - Complete Network Policy support with all features

## Proposal

### Overview

> The proposal will talk about preferred option and also about other options that aims to solve the problem by allowing cloud provider to specify the intent of blocking egress to certain endpoints without blocking all. These deny rules can be plumbed either via iptables or bpf policy rules. IPtables rules would not work with ebpf host routing mode. This proposal considers to solve this via bpf policy rules as it will work for both legacy host routing and ebpf host routing modes. It will also ensure that the behavior of existing network policies or Cilium network policies is not regressed.

Reviewer: nit, there's a lot of words here that are not giving an overview of the solution. This can make it harder for other contributors to understand the proposal.
```suggestion
Allow cloud providers to block egress traffic to certain endpoints without blocking all. This proposal considers to solve this via bpf policy rules as it will work for both legacy host routing and ebpf host routing modes.
```

### Option 1 (CommandLine Flag)

> The decision to pass input as command line was taken based on what is more appropriate for this scenario. There is a section below that describes the reason for going with this option.

Reviewer: nit, I'd suggest describing the options *before* guiding the reader that an option has been pre-selected. This is good context, but it should probably come later (end of subsection) to help describe the motivations behind selecting this approach.

> A new command line flag, --deny-tuples=[ip:port:protocol] could be added as stringSlice. Cloud providers would customize this cilium config to specify apriori that traffic from Cilium-managed Pods towards specific address+port+protocol combinations should be blocked. This flag would effectively act independently of all of the NP, CNP, CCNP resources and enforce a deny for such tuples regardless of the user-defined policy. The deny tuples from cilium config are sanitized and converted to cilium egress deny policy rules. Then these rules should get added to policy map of all endpoints. Several options are considered to implement this and the following section would talk about each option

Reviewer: nit, minor wording about what the proposal is - it's easier for reviewers to understand that this is the concrete plan assuming that this option is taken. Otherwise this is a really good summary of this option.

```suggestion
Add a new command line flag, --deny-tuples=[ip:port:protocol] as stringSlice. Cloud providers would customize this cilium config to specify apriori that traffic from Cilium-managed Pods towards specific address+port+protocol combinations should be blocked. This flag would effectively act independently of all of the NP, CNP, CCNP resources and enforce a deny for such tuples regardless of the user-defined policy. The deny tuples from cilium config are sanitized and converted to cilium egress deny policy rules. Then these rules should get added to policy map of all endpoints. 
```

 * Option 1 is preferred as it aligns with existing code flow and does not affect existing network policies schema or behavior in any way.

#### Adding new field in ruleMetadata

> Based on config received via command line, Cilium Agent would convert it to EgressDeny rule and does policyAdd. PolicyAdd will add the rules to repository and triggers policyreaction event to trigger identity allocation for labels specified in the policy. Cilium by default goes to deny mode if any policy added. To prevent this, this rule can be created with this flag `enableDefaultAllow` enabled. If a rule has this flag enabled, then cilium will never enforce default deny for the rest of traffic unless if there any egress k8s network policy or cilium network policy rule is enforced. By default, this flag is disabled and it would not regress existing network policy scenarios.

Reviewer: > By default, this flag is disabled and it would not regress existing network policy scenarios.

This is a good goal, is there any planned tests to confirm this or do we think the existing testing coverage should give us reliable feedback about this? I guess the existing network policy tests should continue to work even if the new option is enabled, so that would be key to demonstrating this property.

Author: yes will create a separate test with this option enabled and existing network policy should continue to work.

`api.Rule` can extend `ruleMetadata` in `api:rule` structure to add a new field named `enableDefaultAllow`. When evaluating matching policy rules for an endpoint, can differentiate this deny rule from other network policy rules based on this flag. If all matching rules `enableDefaultAllow` to true then default allow will be enforced. If any one of matching rule contains `enableDefaultAllow` to false, then default deny will be enforced.

`Resource` field in `policy.AddOptions` used to indicate the source of policy rule. A new entry of `ResourceKind` named `ResourceKindDaemonDenyTuples` can be added to indicate the source of this rule is via commandLine.

#### Examples
##### Example 1

* Cloud Provider deny tuples 
-–deny-tuples=[“169.254.128.5;80;;]

* Network Policy/Cilium Network Policy 
None

* Action
Egress - Default Allow for all. Block access to 169.254.128.5:80 for all
Ingress - Default Allow for all

##### Example 2

* Cloud Provider deny tuples 

> ```-–d-eny-tuples=[“169.254.128.5;80;;]```

Reviewer: ```suggestion
```-–deny-tuples=[“169.254.128.5;80;;]```
```

* Network Policy/Cilium Network Policy 
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "l3-egress"
spec:
  endpointSelector:
    matchLabels:
      role: backend
  egress:
  - toEndpoints:
    - matchLabels:
        role: frontend
```
* Action
Egress - Allow for all except for pods with label “role: backend”. For those pods, egress allowed only to pods with “role: frontend”. Block access to 169.254.128.5:80 for all
Ingress - Default Allow for all

> #### Alternate Options
> #### Option 2 (CCNP CRD with new field)

Reviewer: nit: If the subsequent options are nested, they should be one step deeper in the document hierarchy. This helps readers to sort through and understand the relationship of this section in comparison to the others. 
```suggestion
### Alternate Options
#### Option 2 (CCNP CRD with new field)
```

> The proposal suggests introducing a new field called "defaultAllowEgress" in the ClusterwideCiliumNetworkPolicy (CCNP). This field aims to enforce a default allow instead of the default deny for egress traffic within the cluster. By default, the defaultAllowEgress field is disabled, indicating that the default deny is in effect. The scope of the defaultAllowEgress field is at the policy level, meaning it applies to the entire policy, even if there are multiple rules defined within the same policy. Default deny for egress traffic will take precedence if egress enabled in atleast one of Network Policy or Cilium Network Policy or CCNP that does not have the defaultAllowEgress flag set. 

Reviewer: ```suggestion
Introduce a new field called "defaultAllowEgress" in the ClusterwideCiliumNetworkPolicy (CCNP). This field aims to enforce a default allow instead of the default deny for egress traffic within the cluster. By default, the defaultAllowEgress field is disabled, indicating that the default deny is in effect. The scope of the defaultAllowEgress field is at the policy level, meaning it applies to the entire policy, even if there are multiple rules defined within the same policy. Default deny for egress traffic will take precedence if egress enabled in at least one of Network Policy or Cilium Network Policy or CCNP that does not have the defaultAllowEgress flag set. 
```

When creating or updating a policy, Cilium triggers the regeneration of the associated endpoint and its policies. It selects all rules that match the endpoint’s identity and in that process, it identifies if egress or ingress policy is enabled. It adds a default allow rule for egress traffic with identity 0 when the egress policy is not enabled. It skips adding this rule when egress policy is enabled. With the proposed change, Cilium will add a new field say  “DefaultAllowEgress” and will be set only if there exists CCNP with defaultAllowEgress set. If there are any other network policies or CCNP without that field, then it will be disabled. With this change, Cilium will add default allow rule for egress traffic if it satisfies any of these two conditions:

- The egress policy is not enabled for the endpoint.
- The egress policy is enabled and the DefaultAllowEgress field is set.

```
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "block-azure-wireserver"
spec:
  endpointSelector: {}
  egressDeny:
  - toCIDR:
    - 169.254.128.5
    toPorts:
    - ports:
      - port: "80"
        protocol: "TCP"
  defaultAllowEgress: true

> ```

Reviewer: This example is presented without context. It's good to see the proposal in context of the policy, but without wording that describes what it does, it's hard to reason about the semantics.

Passing input via Cilium configuration (command-line): Cloud provider policies doesn’t require all features present in other network policies. It doesn't change any existing or any new CRDs and can be kept away from regular users. Additionally, since these policies are not expected to undergo frequent updates, handling them through the Cilium configuration offers a reasonable and manageable approach.

### Key Question: How cloud provider deny rules are processed with network policy rules for same endpoint?
There is no change to default deny behavior of existing network policies even if cloud provider specific deny rules configured via cilium config.  As EgressDeny rules are processed first, deny rules from cloud provider are evaluated first and rest of rules are evaluated later as like other network policy egress deny rules.

### Key Question: Why not iptable rules instead of eBPF?

> cilium adds a rule on top of the FORWARD chain to redirect all traffic to a CILIUM_FORWARD chain. So any Cloud Provider specific firewall rules placed after cilium rules doesn't get executed. Disabling the "prepend-iptables-chains" option is also not a viable solution as disabling prepend means Cilium can no longer guarantee that its policies are enforced. Also in eBPF host routing mode, Cilium bypasses IPtable rules and rules can be enforced only via eBPF. 

Reviewer: nit: These are good context, but for context usually the goal of "Key Questions" is to ask the reviewers to consider the options and provide input on the solutions. These two seem to be more of FAQ items which are probably more suited to the implementatino section.

> ### Considered Option (CommandLine Flag/Cilium Config):

Reviewer: nit:
```suggestion
### Key Question: Which option should be taken to implement the feature?
```
